runners/local-artifact-service-java/src/main/java/org/apache/beam/artifact/local/LocalFileSystemArtifactStagerService.java

```diff
@@ -0,0 +1,215 @@
+       *
+       * <p>Will read starting at the current position of the {@link ByteBuffer} and consume all of
+       * the available bytes.
+       */
+      private void writeData(ByteBuffer byteBuffer) {
+        try {
+          target.write(byteBuffer);
+        } catch (IOException e) {
+          StatusRuntimeException failure = Status.INTERNAL.withCause(e).asRuntimeException();
+          responseObserver.onError(failure);
+          LOG.error("Failed to write contents to file {}", destination, e);
+          throw failure;
+        }
+      }
+
+      @Override
+      public void onError(Throwable t) {
+        try {
+          target.close();
+          // try to cleanup; no guarantees that this will work.
+          destination.delete();
+        } catch (IOException e) {
+          t.addSuppressed(e);
+        }
+      }
+
+      @Override
+      public void onCompleted() {
+        try {
+          target.close();
+        } catch (IOException e) {
+          StatusRuntimeException failure = Status.INTERNAL.withCause(e).asRuntimeException();
+          responseObserver.onError(failure);
+          LOG.error("Failed to complete writing file {}", destination, e);
+          throw failure;
+        }
+        responseObserver.onNext(PutArtifactResponse.getDefaultInstance());
+        responseObserver.onCompleted();
+      }
+    };
+  }
+
+  @Override
+  public void commitManifest(
+      CommitManifestRequest request, StreamObserver<CommitManifestResponse> responseObserver) {
+    Collection<Artifact> missing = new ArrayList<>();
+    for (Artifact artifact : request.getManifest().getArtifactList()) {
+      if (!getArtifactFile(artifact.getName()).exists()) {
+        missing.add(artifact);
+      }
+    }
+    if (missing.isEmpty()) {
+      File mf = new File(stagingBase, "MANIFEST");
+      try {
+        checkState(mf.createNewFile());
+        try (OutputStream mfOut = new FileOutputStream(mf)) {
+          request.getManifest().writeTo(mfOut);
+        }
+        responseObserver.onNext(
+            CommitManifestResponse.newBuilder()
+                .setStagingToken(stagingBase.getCanonicalPath())
+                .build());
+        responseObserver.onCompleted();
+      } catch (IOException e) {
+        StatusRuntimeException failure = Status.INTERNAL.withCause(e).asRuntimeException();
+        responseObserver.onError(failure);
+        LOG.error("Failed to commit Manifest {}", request.getManifest(), e);
+        throw failure;
+      }
+    } else {
+      StatusRuntimeException failure =
+          Status.INVALID_ARGUMENT
+              .withDescription(
+                  String.format(
+                      "Attempted to commit manifest with missing Artifacts: [%s]", missing))
+              .asRuntimeException();
+      responseObserver.onError(failure);
+      throw failure;
+    }
+  }
+
+  private File getArtifactFile(String artifactName) {
+    // TODO: Create a subdirectory rather than adding a prefix
```

Comment:
<!--new_thread; commit:fd3270d35ab3f8121ad44c2521dd0ade4223d84c; resolved:0-->
Is there a good reason not to do this? I was just writing up a comment on your placement of MANIFEST when I noticed this.

Reply:
Done.